pkg/consistent/connection.go

```diff
@@ -0,0 +1,262 @@
+	GetFailedTime() int64
+	NeedGC(interval time.Duration) bool
+}
+
+type subConnWrapper struct {
+	cc           balancer.ClientConn
+	healthCheck  bool
+	sc           balancer.SubConn
+	addr         string
+	state        connectivity.State
+	lastKeepTime time.Time
+	useCount     *atomic.Int64
+	failCount    int64 // no need to be thread-safe
+}
+
+func NewSubConnWrapper(conn balancer.ClientConn, addr string, healthCheck bool) (SubConnWrapper, error) {
+	scw := &subConnWrapper{
+		cc:           conn,
+		healthCheck:  healthCheck,
+		addr:         addr,
+		lastKeepTime: time.Now(),
+		useCount:     atomic.NewInt64(0),
+	}
+	sc, err := scw.cc.NewSubConn([]resolver.Address{{Addr: addr}}, balancer.NewSubConnOptions{HealthCheckEnabled: healthCheck})
+	if err != nil {
+		return nil, err
+	}
+	scw.sc = sc
+	scw.state = connectivity.Idle
+	return scw, nil
+}
+
+func (scw *subConnWrapper) GetConn() balancer.SubConn {
+	return scw.sc
+}
+
+func (scw *subConnWrapper) Address() string {
+	return scw.addr
+}
+
+func (scw *subConnWrapper) Keep() {
+	scw.lastKeepTime = time.Now()
+}
+
+func (scw *subConnWrapper) State() connectivity.State {
+	return scw.state
+}
+
+func (scw *subConnWrapper) AddUsingCount(n int64) {
+	scw.useCount.Add(n)
+}
+
+func (scw *subConnWrapper) GetFailedTime() int64 {
+	return scw.failCount
+}
+
+func (scw *subConnWrapper) NeedGC(interval time.Duration) bool {
+	if scw.State() != connectivity.Ready {
+		return false
+	}
+	if scw.useCount.Load() > 0 {
+		return false
+	}
+	if scw.lastKeepTime.Add(interval).Before(time.Now()) {
+		return true
+	}
+	return false
+}
+
+func (scw *subConnWrapper) UpdateState(state connectivity.State) {
+	scw.state = state
+	if state == connectivity.TransientFailure {
+		scw.failCount++
+	}
+	if state == connectivity.Ready {
+		scw.failCount = 0
+	}
+}
+
+func (scw *subConnWrapper) Connect() {
+	scw.sc.Connect()
+	scw.Keep()
+}
+
+func (scw *subConnWrapper) Disconnect() {
+	scw.cc.RemoveSubConn(scw.sc)
+}
+
+func (scw *subConnWrapper) Close() {
+	scw.cc.RemoveSubConn(scw.sc)
+}
+
+type conPool struct {
+	*sync.RWMutex
+	cons       map[string]SubConnWrapper
+	subCons    map[balancer.SubConn]SubConnWrapper
+	rand       *rand.Rand
+	gcInterval time.Duration
+	exit       chan struct{}
+}
+
+func NewConnPool(gcInterval time.Duration) ConnPool {
+	cp := &conPool{
+		RWMutex:    &sync.RWMutex{},
+		cons:       map[string]SubConnWrapper{},
+		subCons:    map[balancer.SubConn]SubConnWrapper{},
+		rand:       rand.New(rand.NewSource(time.Now().UnixNano())),
+		gcInterval: gcInterval,
+		exit:       make(chan struct{}, 1),
+	}
+	if gcInterval != 0 {
+		go cp.Watch()
+	}
+	return cp
+}
+
+func (cp *conPool) Watch() {
+	ticker := time.NewTicker(time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			for _, scw := range cp.Values() {
+				if scw.NeedGC(cp.gcInterval) {
+					scw.UpdateState(TemporarilyClosed)
+					scw.Disconnect()
```

Comment:
Should also remove sc in conPool.cons and conPool.subCons ?

Reply:
remove operation is done in `balancer.UpdateSubConnState` when subConn state become ShutDown